### PR TITLE
Fix `wp_head` performance regression for classic themes

### PR DESCRIPTION
**Pull request by felixarntz**
* A performance analysis for WordPress 6.1 has shown that there is a performance regression in `wp_head` due to additional logic focused on FSE, which particularly for classic themes appears to be unnecessary.
* See relevant Trac ticket comments discussing this so far:
    * https://core.trac.wordpress.org/ticket/56467#comment:408
    * https://core.trac.wordpress.org/ticket/56467#comment:411
    * https://core.trac.wordpress.org/ticket/56467#comment:413
* The original implementation of this PR improves median `wp_head` execution time from **16.45ms to 11.92ms (~28% faster)** for classic themes (concretely tested with Twenty Twenty-One), while apparently not missing any actual functionality needed.
* However, based on https://github.com/WordPress/wordpress-develop/pull/3536#issuecomment-1307699049, a more complex part of this implementation was broken out into a separate Gutenberg PR https://github.com/WordPress/gutenberg/issues/45616, so that the rest of this PR can realistically make it into the WordPress 6.1.1 release.
* The latest implementation of this PR still improves median `wp_head` execution time from **16.45ms to 14.78ms (~10% faster)** for the same test environment. This is quite a bit less of an improvement from the original, but it makes sense as with the latest implementation still core's `theme.json` and `theme-i18n.json` files are being parsed. That said, this still brings a worthwhile performance improvement and sets the stage for the follow up work in https://github.com/WordPress/gutenberg/issues/45616 to address the remaining part of the performance issues.
* See [this spreadsheet](https://docs.google.com/spreadsheets/d/1q6PNZ-7uIPJ9MTC3JiXshpNyG7sm28QJOtmavKexsIU/edit) for the test results.

cc @scruffian @oandregal @draganescu 

Trac ticket: https://core.trac.wordpress.org/ticket/56945

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 1]
```diff
@@ -170,8 +170,12 @@ public static function get_core_data() {
 			return static::$core;
 		}
 
-		$config = static::read_json_file( __DIR__ . '/theme.json' );
-		$config = static::translate( $config );
+		if ( static::theme_has_support() ) {
```
Comment (eugene-manuilov) [inline src/wp-includes/class-wp-theme-json-resolver.php]: This is not relevant for this PR but I have also found that we can remove redundant `is_readable` function calls in the `theme_has_support` method.

Currently we call it once or twice for results returned from the `get_file_path_from_theme` method: https://github.com/WordPress/wordpress-develop/blob/aff634740769140f5f9f2961ba741c544ee92f57/src/wp-includes/class-wp-theme-json-resolver.php#L608-L611

But the `get_file_path_from_theme` method suddenly also calls the same `is_readable` function before returning non-empty result: https://github.com/WordPress/wordpress-develop/blob/aff634740769140f5f9f2961ba741c544ee92f57/src/wp-includes/class-wp-theme-json-resolver.php#L633

So, we can update the `theme_has_support` method to check whether the returned result from the `get_file_path_from_theme` is not empty: 	
```diff
	public static function theme_has_support() {
		if ( ! isset( static::$theme_has_support ) ) {
			static::$theme_has_support = (
-				is_readable( static::get_file_path_from_theme( 'theme.json' ) ) ||
-				is_readable( static::get_file_path_from_theme( 'theme.json', true ) )
+				static::get_file_path_from_theme( 'theme.json' ) !== '' ||
+				static::get_file_path_from_theme( 'theme.json', true ) !== ''
			);
		}

		return static::$theme_has_support;
	}
```

P.S.: calling the `is_readable` function in the `get_file_path_from_theme` method is a controversial decision which is not expected from that function and honestly shouldn't be there IMO.

**Comment by spacedmonkey** [comment]
cc @oandregal

**Comment by oandregal** [comment]
:wave: I've seen the changes and this is what I can share:

- `get_core_data`: this data needs to be available both for classic and block themes. It loads the default editor configuration (for example, the default color palettes)
- `get_block_data`: this data is necessary for both classic and block themes. It loads style information from blocks.
- `get_theme_data`: this data is required for both classic and block themes. For classic, it takes it from the theme supports and returns it as if it was a theme.json. There may be some opportunities to micro-optimize here, such as do not load the i18n schema if the theme and its parent do not have a theme.json (we'd need to measure how impactful that single change is, that´d be a nice single PR).
- `get_user_data`: this is safe to skip for classic themes. There are edge cases, such as a theme that has theme.json and then it removes it, though I'm inclined to think user styles are safe to ignore in that case. Coincidentally, I've done the same thing in a [prototype](https://github.com/WordPress/gutenberg/pull/45372) I'm working on. It can be extracted as a small, focused PR.

Coincidentally, for the area of "theme.json APIs", I'm compiling a list of tasks I find most pressing at https://github.com/WordPress/gutenberg/issues/45171 I'm very optimistic about how those will impact the performance and code quality. For example, I've just published a [prototype](https://github.com/WordPress/gutenberg/pull/45372) that reduces the total time of a request from 514.44ms to 343.71ms, a 33% performance increase.

Another thing that it'd be useful is to prepare these PRs directly in Gutenberg. That way, they can be tested widely and bugs will be raised earlier than the WordPress RC cycles, which is always a lot less stressful and give people time to figure out long-term solutions rather than patches for the RC.

**Comment by oandregal** [comment]
> This PR improves median wp_head execution time from 16.45ms to 11.92ms (~28% faster) for classic themes (concretely tested with Twenty Twenty-One

Felix, would you share the steps you did to test this, so it can be replicated in a variety of environments?

As a side note, it'd be great to track some statistics across WordPress versions and watch its evolution, like we do for the editor (see live at https://codehealth.vercel.app/).

**Comment by felixarntz** [comment]
@oandregal Thank you for providing additional context on why (parts of) these methods are needed for classic themes as well. Based on what you're saying, I still think there are a number of performance optimizations to make here to avoid running FSE-only logic for classic themes. Here are some of the high-level ideas I'm thinking:
* In `get_core_data()`, parsing WP's own `theme.json` is a rather wasteful implementation in itself, more so with the additional translation parsing. Since this file is shipped by core (and thus never changes between core updates), we should find a more efficient solution. For example, how about we generate a fully PHP-ready parsed version including i18n strings of WP's `theme.json` at build time? I see how `theme.json` is certainly more portable than the same data in PHP array format, but we could create an autogenerated version of that file, like a `theme-json.php` file that can then simply be `require`d by PHP, no JSON parsing and i18n schema parsing necessary at runtime.
* In `get_block_data()`, everything looks fine as is to me.
* In `get_theme_data()`, we should simply not parse the i18n schema and try to parse the theme's `theme.json` if it doesn't exist, like you're saying and like this PR already does.
* `get_user_data()` can bail early if in a classic theme, also like you're saying and like this PR already does.

I'll make some updates to the proposed code here accordingly, the main thing though that requires some more work would be to generate some version of a `theme-json.php` file as part of the build process. If we build something like that, we could even share it with e.g. theme developers so they can run it on their `theme.json` files too.

> Felix, would you share the steps you did to test this, so it can be replicated in a variety of environments?

It was a fairly basic process, here's what I did:
* Modify WP core's `wp_head()` function directly, to measure `microtime( true )` before the `do_action( 'wp_head' )` call and after it, then get the difference to know how long the action took.
* I did this with `trunk` and with this PR branch, using a site with Twenty Twenty-One on just the regular basic "Hello world" singular post URL.
* I ran both tests 9 times each (to account for variance) and calculated the median.

If we wanted to dive in more deeply, it would also be interesting to see what the difference would be on a more complex piece of content, e.g. the Gutenberg demo page.

**Comment by felixarntz** [comment]
@oandregal I've pushed some updates here as mentioned above, please have a look. I have included what the `theme-json.php` file could look like in principle, although of course this would need to be auto-generated by a build process eventually.

One additional concern/question I have: Why is the condition around whether to reuse the previously "cached" theme data only present around parsing the (child) theme's `theme.json` (see https://github.com/WordPress/wordpress-develop/pull/3536/files#diff-d6b86476eed058e7cf8b6e57fa52c4fd75b1f0907e1a9ccb0149528a24f7578bR254)? Are there any problems if we change it to apply to the entire method, similar to how it is for the other `get_*_data()` methods? Otherwise some of that logic is still run on every single method call over and over again. cc @hellofromtonya

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 1 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php]: @eugene-manuilov I've included this fix here.

**Comment by spacedmonkey** [comment]
This PR is heavily related - https://github.com/WordPress/wordpress-develop/pull/3556.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 2]
```diff
@@ -263,16 +273,19 @@ public static function get_theme_data( $deprecated = array(), $options = array()
...
-			$parent_theme           = new WP_Theme_JSON( $parent_theme_json_data );
-
-			/*
-			 * Merge the child theme.json into the parent theme.json.
-			 * The child theme takes precedence over the parent.
-			 */
-			$parent_theme->merge( static::$theme );
-			static::$theme = $parent_theme;
+			$parent_theme_json_file = static::get_file_path_from_theme( 'theme.json', true );
+			if ( $parent_theme_json_file !== '' ) {
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-theme-json-resolver.php]: Yoda case we need.

File: src/wp-includes/class-wp-theme-json.php [hunk 3]
```diff
@@ -2934,7 +2934,8 @@ public function get_data() {
 	public function set_spacing_sizes() {
 		$spacing_scale = _wp_array_get( $this->theme_json, array( 'settings', 'spacing', 'spacingScale' ), array() );
 
-		if ( ! is_numeric( $spacing_scale['steps'] )
+		if ( ! isset( $spacing_scale['steps'] )
+			|| ! is_numeric( $spacing_scale['steps'] )
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-theme-json.php:2943]: Why is this change needed?

**Comment by spacedmonkey** [comment]
This is hard to code review the code to generate theme-php.php, fix the unit test and coding standards.

**Comment by felixarntz** [comment]
@spacedmonkey 
> This is hard to code review the code to generate theme-php.php, fix the unit test and coding standards.

No point to review the `theme-json.php` file, as eventually it will be auto-generated. There's also no point in fixing the WPCS etc violations in that file - see the TODO comment and the issue comments above outlining that. :)

I'll work on further polishing here today.

@draganescu @hellofromtonya Any chance I could get your thoughts on the changes and proposed approach from this draft PR? Please make sure to review my above issue comments for context.

**Comment by spacedmonkey** [comment]
>No point to review the theme-json.php file, as eventually it will be auto-generated. There's also no point in fixing the WPCS etc violations in that file - see the TODO comment and the issue comments above outlining that. :)

I don't think we should merge this without the automatic generation code, in JS grunt. 

I think we should consider making this out to another PR,  where we have more time to review / test.

**Comment by felixarntz** [comment]
@spacedmonkey 
> I don't think we should merge this without the automatic generation code, in JS grunt.

That's exactly what I'm saying.

> I think we should consider making this out to another PR, where we have more time to review / test.

Sounds good, I've opened https://github.com/WordPress/gutenberg/issues/45616 for that.

Will remove the `theme-json.php` part from here and get this PR ready for 6.1.1.

File: src/wp-includes/class-wp-theme-json.php (hunk 3 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php:2943]: @spacedmonkey This is mostly about defensive coding. When a parsed `WP_Theme_JSON` has empty data, without the `! isset()` here you get a PHP warning.

**Comment by felixarntz** [comment]
@spacedmonkey @draganescu @hellofromtonya This PR is now ready for a full review, including test coverage. Would be great to get your feedback so that we can aim to get this into 6.1.1.

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 4]
```diff
@@ -681,11 +685,36 @@ function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries_
...
+	 * @covers WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles
+	 */
+	function test_get_user_data_from_wp_global_styles_does_not_run_for_theme_without_support() {
+		// The 'default' theme does not support theme.json.
+		switch_theme( 'default' );
+		wp_set_current_user( self::$administrator_id );
+		$theme = wp_get_theme();
+
+		// Add filter to record DB query count.
+		add_filter( 'query', array( $this, 'filter_db_query' ) );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]:
```suggestion
		$start_queries = get_num_queries();
```

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 5]
```diff
@@ -681,11 +685,36 @@ function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries_
...
+
+		// When theme.json is not supported, the method should not run a query and always return an empty result.
+		$user_cpt = WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles( $theme );
+		$this->assertEmpty( $user_cpt, 'User CPT is expected to be empty.' );
+		$this->assertSame( 0, count( $this->queries ), 'Unexpected SQL query detected for theme without theme.json support.' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]:
```suggestion
		$this->assertSame( 0, get_num_queries() - $start_queries ), 'Unexpected SQL query detected for theme without theme.json support.' );
```

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 6]
```diff
@@ -681,11 +685,36 @@ function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries_
...
+
+		$user_cpt = WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles( $theme, true );
+		$this->assertEmpty( $user_cpt, 'User CPT is expected to be empty.' );
+		$this->assertSame( 0, count( $this->queries ), 'Unexpected SQL query detected for theme without theme.json support.' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]:
```suggestion
		$this->assertSame( 0, get_num_queries() - $start_queries, 'Unexpected SQL query detected for theme without theme.json support.' );
```

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 7]
```diff
@@ -582,8 +593,8 @@ public static function get_user_global_styles_post_id() {
 	public static function theme_has_support() {
 		if ( ! isset( static::$theme_has_support ) ) {
 			static::$theme_has_support = (
-				is_readable( static::get_file_path_from_theme( 'theme.json' ) ) ||
-				is_readable( static::get_file_path_from_theme( 'theme.json', true ) )
+				static::get_file_path_from_theme( 'theme.json' ) !== '' ||
```
Comment (oandregal) [inline src/wp-includes/class-wp-theme-json-resolver.php:605]: This is nice, `get_file_path_from_theme` already calls `is_readable`.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 8]
```diff
@@ -582,8 +593,8 @@ public static function get_user_global_styles_post_id() {
...
+				static::get_file_path_from_theme( 'theme.json', true ) !== ''
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-theme-json-resolver.php:606]: Yoda

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 9]
```diff
@@ -259,20 +264,23 @@ public static function get_theme_data( $deprecated = array(), $options = array()
...
+				$parent_theme_json_file = static::get_file_path_from_theme( 'theme.json', true );
+				if ( '' !== $parent_theme_json_file ) {
+					$parent_theme_json_data = static::read_json_file( $parent_theme_json_file );
+					$parent_theme_json_data = static::translate( $parent_theme_json_data, wp_get_theme()->parent()->get( 'TextDomain' ) );
+					$parent_theme           = new WP_Theme_JSON( $parent_theme_json_data );
+
+					/*
+					* Merge the child theme.json into the parent theme.json.
+					* The child theme takes precedence over the parent.
+					*/
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-theme-json-resolver.php]:
```suggestion
					 * Merge the child theme.json into the parent theme.json.
					 * The child theme takes precedence over the parent.
					 */
```

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 10]
```diff
@@ -400,6 +408,9 @@ private static function remove_json_comments( $array ) {
 	 * @return array Custom Post Type for the user's origin config.
 	 */
 	public static function get_user_data_from_wp_global_styles( $theme, $create_post = false, $post_status_filter = array( 'publish' ) ) {
+		if ( ! static::theme_has_support() ) {
+			return array();
+		}
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-theme-json-resolver.php]: The order seems incorrect here. The check for the theme support happens before `$theme` is confirmed as an instance of `WP_Theme`.

If that's correct, are you able to jot down why in a comment?

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 11]
```diff
@@ -634,6 +634,8 @@ function test_merges_child_theme_json_into_parent_theme_json() {
 	 * @covers WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles
 	 */
 	function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries() {
+		// Switch to a theme that does have support.
+		switch_theme( 'block-theme' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php:624]: Fixing tests always makes me nervous, why is this needed now? Maybe expand the comment so others can understand too. 

:1234: This applies to the other tests with the same change.

**Review by peterwilsoncc — COMMENTED** [review]
Added a fwe notes inline.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 12]
```diff
@@ -400,6 +408,9 @@ private static function remove_json_comments( $array ) {
 	 * @return array Custom Post Type for the user's origin config.
 	 */
 	public static function get_user_data_from_wp_global_styles( $theme, $create_post = false, $post_status_filter = array( 'publish' ) ) {
+		if ( ! static::theme_has_support() ) {
```
Comment (oandregal) [inline src/wp-includes/class-wp-theme-json-resolver.php]: The only edge case I can think of is a theme that uses `theme.json` and removes it in a later version. In that case, this changes the behavior: the code will now ignore user data while it was used before.

**Comment by oandregal** [comment]
:wave: I've run some local profiling tests using TwentyTwentyOne. My test was loading the "hello world" post as a logged-out user. These are the numbers (median of 9 tries): it takes 393.98ms with this request and 411.04ms without these changes. This improves 17.06ms, a 4.15% improvement for that theme.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 13]
```diff
@@ -246,8 +246,13 @@ public static function get_theme_data( $deprecated = array(), $options = array()
 		$options = wp_parse_args( $options, array( 'with_supports' => true ) );
 
 		if ( null === static::$theme || ! static::has_same_registered_blocks( 'theme' ) ) {
-			$theme_json_data = static::read_json_file( static::get_file_path_from_theme( 'theme.json' ) );
-			$theme_json_data = static::translate( $theme_json_data, wp_get_theme()->get( 'TextDomain' ) );
+			$theme_json_file = static::get_file_path_from_theme( 'theme.json' );
```
Comment (oandregal) [inline src/wp-includes/class-wp-theme-json-resolver.php:249]: This change is fine. I can't think of a reason why this would break existing behavior.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 14]
```diff
@@ -259,20 +264,23 @@ public static function get_theme_data( $deprecated = array(), $options = array()
...
-			/*
-			 * Merge the child theme.json into the parent theme.json.
-			 * The child theme takes precedence over the parent.
-			 */
-			$parent_theme->merge( static::$theme );
-			static::$theme = $parent_theme;
+			if ( wp_get_theme()->parent() ) {
+				// Get parent theme.json.
+				$parent_theme_json_file = static::get_file_path_from_theme( 'theme.json', true );
+				if ( '' !== $parent_theme_json_file ) {
```
Comment (oandregal) [inline src/wp-includes/class-wp-theme-json-resolver.php:271]: This change is fine as well.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 15]
```diff
@@ -259,20 +264,23 @@ public static function get_theme_data( $deprecated = array(), $options = array()
 			$theme_json      = apply_filters( 'wp_theme_json_data_theme', new WP_Theme_JSON_Data( $theme_json_data, 'theme' ) );
 			$theme_json_data = $theme_json->get_data();
 			static::$theme   = new WP_Theme_JSON( $theme_json_data );
-		}
```
Comment (oandregal) [inline src/wp-includes/class-wp-theme-json-resolver.php:262]: This looks like an oversight. I can't think of a reason this would have been desired (check for the parent even if the child had already been computed and blocks didn't change).

**Comment by oandregal** [comment]
> This improves 17.06ms, a 4.15% improvement for that theme.

For reference, block themes also improve in 12.47ms.

This PR does a few changes:

[1](https://github.com/WordPress/wordpress-develop/pull/3536/files#r1017426094). remove `is_readable` for theme.json check (both themes with and without `theme.json`)
[2](https://github.com/WordPress/wordpress-develop/pull/3536/files#r1017485880). avoid looking up the parent if the theme data is already set (both themes with and withouth `theme.json`)
[3](https://github.com/WordPress/wordpress-develop/pull/3536/files#r1017444637). remove one query for themes without `theme.json`
[4](https://github.com/WordPress/wordpress-develop/pull/3536/files#r1017480613). do not load the translation for themes with no `theme.json`

I've done my testing with TwentyTwentyOne and TwentyTwentyThree (none is a child theme). Given the changes and the numbers are so close (-12ms vs -17ms for themes with and without `theme.json`), it sounds like the most impactful change was actually 1 (the only code that both themes have in common in my testing). Removing the query for classic themes is a 5ms improvement at best. I'm time-bounded at the moment and could not dig deeper. My impression is that the results are not lined up with the initial expectations (we are doing _more_ work for classic themes). Would that be a fair statement?

I'm a bit torn about whether this is worth doing. It'd be great if we could see if only applying the changes at 1 and 2 brings the same amount of improvement. If that's the case I'd remove all the other changes, as they make the code harder to read and would not bring a gain worth pursuing.

As I said, I won't have much time to look into this in the following days, though it looks like a safe change from a code point of view. Hope this is helpful.

**Comment by oandregal** [comment]
I also want to share the two things that I consider more impactful at the moment to focus on:

1) Review and [cache the public API methods](https://github.com/WordPress/gutenberg/issues/45171). An initial investigation I did resulted in [this PR](https://github.com/WordPress/gutenberg/pull/45372), which provides a 200ms reduction, a 33% improvement. Every ms counts, so squeezing all of them at the maximum is always helpful. Though there are so many things to work on that I think it'd be helpful to look at performance strategically: intervene for the bigger wins first, then look at the smaller bits. I'll continue to look into this more when I'm back at full speed.

2) Having an automated way to look at numbers and define what's important for us. At the moment, everyone is tracking things manually, and it takes a lot of time from the PR author and from the reviewers. It's also a bit inconsistent: should we measure the full time-to-request as logged-out users or is it best tracking particular actions such as `wp_head`? Having a common understanding of what to improve and tracking automatically instead of manually is a huge win that would help us to move forward faster. Unfortunately, I don't think I'll be able to help in this area much. It'd be lovely if people with more experience and background in setting these things up would take the lead.

**Comment by spacedmonkey** [comment]
In trunk
Screenshot 2022-11-09 at 10 50 57
In this branch. 
Screenshot 2022-11-09 at 10 50 13

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 12 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-theme-json-resolver.php]: YES! Love this change.

**Review by spacedmonkey — APPROVED** [review]
PR works. It doesn't solve all the issues in wp_head, but it a great start

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php (hunk 4 above)
Comment (felixarntz) [reply tests/phpunit/tests/theme/wpThemeJsonResolver.php]: Good idea, updated. Also in the other pre-existing cases in this tests file.

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php (hunk 5 above)
Comment (felixarntz) [reply tests/phpunit/tests/theme/wpThemeJsonResolver.php]: See above, updated throughout.

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php (hunk 6 above)
Comment (felixarntz) [reply tests/phpunit/tests/theme/wpThemeJsonResolver.php]: Updated.

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 8 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php:606]: I don't think this is needed here. Yoda conditions are only required around variable to value comparisons, not with functions (since setting a function to another value would result in an error anyway).

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 9 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php]: Good catch, updated.

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 10 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php]: Good pointer, although the problem is in fact slightly different here. The order doesn't matter as much as the fact that `theme_has_support()` only supports the currently active theme, however the `$theme` parameter could be any theme (although in all use-cases in production code and even the tests, it _is_ the currently active theme).

I've therefore updated the logic here to only short-circuit if `$theme` is in fact the active theme. Preferably we should eventually add support for any theme to the `theme_has_support()` method to make this more stable, although one could also argue that's kind of pointless since I can't see a situation where you would want to call this for a non-active theme.

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php (hunk 11 above)
Comment (felixarntz) [reply tests/phpunit/tests/theme/wpThemeJsonResolver.php:624]: Previously the logic here simply ignored the fact that classic themes don't need to run this logic, therefore the tests didn't bother switching to a block theme. Fixing this is precisely part of the goals of this PR, therefore the other tests that were operating under the assumption of a block theme now need to be updated to actually enforce a block theme.

So I'd argue it was rather the existing tests here that were previously too vague.

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 12 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php]: That's a fair point. However, I would consider this indeed an edge-case which I don't think is worth addressing here. If a theme removed its `theme.json` it would come with a whole lot of other behavioral changes anyway - I don't think realistically any theme author is going to change a block theme to become a classic theme.

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 15 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php:262]: Thanks, makes sense. It looked like an oversight to me as well.

**Comment by felixarntz** [comment]
@oandregal Thank you for running additional performance benchmarks on the latest code here! This is super helpful.

I've also run the latest iteration of the PR through the same methodology that I originally used, and posted results in an updated PR description. Of course the benefit is much less than originally, now that I've broken out fixing the core `theme.json` parsing piece into https://github.com/WordPress/gutenberg/issues/45616. Nevertheless, at a ~10% `wp_head` performance improvement I think this is a worthwhile change especially as they look straightforward from a code perspective - baby steps so to speak, but together they'll add up eventually, and doing the work here is particularly relevant to allow for the above Gutenberg issue to actually have the intended impact.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 16]
```diff
@@ -408,12 +408,16 @@ private static function remove_json_comments( $array ) {
...
-		if ( ! static::theme_has_support() ) {
-			return array();
-		}
 		if ( ! $theme instanceof WP_Theme ) {
 			$theme = wp_get_theme();
 		}
+
+		// Bail early if the theme does not support a theme.json. Since WP_Theme_JSON_Resolver::theme_has_support()
+		// only supports the active theme, the extra condition for whether $theme is the active theme is present here.
+		if ( $theme->get_stylesheet() === get_stylesheet() && ! static::theme_has_support() ) {
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-theme-json-resolver.php:422]: Instead of doing this, how about we add a method to the WP_Theme class. So we could do this, 
```
if( $theme->theme_has_support() ) {
```

We could even use the `get_file_path` method of the WP_Theme class.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 17]
```diff
@@ -403,6 +411,13 @@ public static function get_user_data_from_wp_global_styles( $theme, $create_post
 		if ( ! $theme instanceof WP_Theme ) {
 			$theme = wp_get_theme();
 		}
+
+		// Bail early if the theme does not support a theme.json. Since WP_Theme_JSON_Resolver::theme_has_support()
+		// only supports the active theme, the extra condition for whether $theme is the active theme is present here.
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-theme-json-resolver.php]: Nits:
* multiline comment formatting
* I think we target commment lenghts of 80chars, so I've reformatted the lines too.

```suggestion
		/*
		 * Bail early if the theme does not support a theme.json.
		 *
		 * Since WP_Theme_JSON_Resolver::theme_has_support() only supports the active
		 * theme, the extra condition for whether $theme is the active theme is
		 * present here.
		 */
```

**Review by peterwilsoncc — COMMENTED** [review]
LGTM with a nit pick inline

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 16 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-theme-json-resolver.php:422]: I have worked on a moving this to the WP_Theme class. See https://github.com/spacedmonkey/wordpress-develop/commit/b10228fc4aa106438e0a989f30b7340973d8f786. This change is inline with `is_block_theme` and `is_block_theme` method on WP_theme.

Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php:422]: @spacedmonkey I like that idea, however I think this goes beyond the scope of what would fit into a WordPress patch release. I think adding API methods/functions should be part of a regular WP release like 6.2.

Comment (spacedmonkey) [reply src/wp-includes/class-wp-theme-json-resolver.php:422]: The problem is, I don't love this line. 

```
if ( $theme->get_stylesheet() === get_stylesheet() && ! static::theme_has_support() ) {
```

It is only relates to current theme. My change would fix that issue.

Comment (spacedmonkey) [reply src/wp-includes/class-wp-theme-json-resolver.php:422]: At the very least, we should not use wp_get_theme that creates a new object each time.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 18]
```diff
@@ -259,20 +264,23 @@ public static function get_theme_data( $deprecated = array(), $options = array()
...
-			$parent_theme_json_data = static::translate( $parent_theme_json_data, wp_get_theme()->parent()->get( 'TextDomain' ) );
-			$parent_theme           = new WP_Theme_JSON( $parent_theme_json_data );
 
-			/*
-			 * Merge the child theme.json into the parent theme.json.
-			 * The child theme takes precedence over the parent.
-			 */
-			$parent_theme->merge( static::$theme );
-			static::$theme = $parent_theme;
+			if ( wp_get_theme()->parent() ) {
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-theme-json-resolver.php:268]: This method, uses `wp_get_theme()` ALOT. This creates a new WP_Theme instance everytime. Settings a variable like this `$wp_theme = wp_get_theme();` would have some CPU cycles.

Comment (spacedmonkey) [reply src/wp-includes/class-wp-theme-json-resolver.php:268]: Created a follow up issue - https://core.trac.wordpress.org/ticket/57057

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 16 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php:422]: I saw you opened a follow up ticket, SGTM!

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 18 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json-resolver.php:268]: Thanks, will review it today!

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/54799.
